include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -0,0 +1,231 @@
+		const bool valid = irr_glsl_calcAnisotropicMicrofacetCache(uf, true, currInteraction.isotropic.V.dir, s.L, currInteraction.T, currInteraction.B, currInteraction.isotropic.N, s.NdotL, s.VdotL, orientedEta, rcpOrientedEta);
+		//assert(valid);
+	}
+	else
+	{
+		uf = _uf;
+	}
+
+	if (cosFactor>FLT_MIN && op_hasSpecular(op))
+	{
+		BEGIN_CASES(ndf)
+#ifdef NDF_GGX
+		CASE_BEGIN(ndf, NDF_GGX) {
+
+#if defined(OP_THINDIELECTRIC) || defined(OP_DIELECTRIC)
+			if (is_bsdf) {
+#ifdef ALL_ISOTROPIC_BXDFS
+				bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_dielectric_cos_eval(s, currInteraction.isotropic, uf.isotropic, ior[0].x, a2);
+#else
+				bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_dielectric_cos_eval(s, currInteraction.isotropic, uf.isotropic, ior[0].x, a2);
+#endif
+			}
+			else
+#endif
+			{
+#ifdef ALL_ISOTROPIC_BXDFS
+				bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_cos_eval_DG(s, uf.isotropic, currInteraction.isotropic, a2);
+#else
+				bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(s, uf, currInteraction, a, a2, ay, ay2);
+#endif
+			}
+
+		} CASE_END
+#endif
+
+#ifdef NDF_BECKMANN
+		CASE_BEGIN(ndf, NDF_BECKMANN) {
+
+#if defined(OP_THINDIELECTRIC) || defined(OP_DIELECTRIC)
+			if (is_bsdf) {
+#ifdef ALL_ISOTROPIC_BXDFS
+				bxdf_eval_scalar_part = irr_glsl_beckmann_smith_height_correlated_dielectric_cos_eval_wo_cache_validation(s, currInteraction, uf, ior[0].x, a2);
+#else
+				bxdf_eval_scalar_part = irr_glsl_beckmann_aniso_smith_height_correlated_cos_eval_wo_cache_validation(s, currInteraction, uf, ior[0].x, a, a2, ay, ay2);
+#endif
+			}
+			else
+#endif
+			{
+#ifdef ALL_ISOTROPIC_BXDFS
+				bxdf_eval_scalar_part = irr_glsl_beckmann_height_correlated_cos_eval_DG(s, uf.isotropic, currInteraction.isotropic, a2);
+#else
+				bxdf_eval_scalar_part = irr_glsl_beckmann_aniso_height_correlated_cos_eval_DG(s, uf, currInteraction, a, a2, ay, ay2);
+#endif
+			}
+
+		} CASE_END
+#endif
+
+#ifdef NDF_PHONG
+		CASE_BEGIN(ndf, NDF_PHONG) {
+
+#if defined(OP_THINDIELECTRIC) || defined(OP_DIELECTRIC)
+			if (is_bsdf) {
+#ifdef ALL_ISOTROPIC_BXDFS
+				bxdf_eval_scalar_part = irr_glsl_beckmann_smith_height_correlated_dielectric_cos_eval_wo_cache_validation(s, currInteraction, uf, ior[0].x, a2);
+#else
+				bxdf_eval_scalar_part = irr_glsl_beckmann_aniso_smith_height_correlated_cos_eval_wo_cache_validation(s, currInteraction, uf, ior[0].x, a, a2, ay, ay2);
+#endif
+			}
+			else
+#endif
+			{
+				float n = irr_glsl_alpha2_to_phong_exp(a2);
+#ifdef ALL_ISOTROPIC_BXDFS
+				bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(s, uf.isotropic, currInteraction.isotropic, n, a2);
+#else
+				float ny = irr_glsl_alpha2_to_phong_exp(ay2);
+				bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(s, uf, currInteraction, n, ny, a2, ay2);
+#endif
+			}
+
+		} CASE_END
+#endif
+
+		CASE_OTHERWISE
+		{} //else "empty braces"
+		END_CASES
+	}
```

Comment:
same comment as here
https://github.com/buildaworldnet/IrrlichtBAW/pull/424#discussion_r509628370

https://github.com/buildaworldnet/IrrlichtBAW/pull/424#discussion_r509626472

Reply:
basically bad if statement, could de-diverge dielectric and conductor, etc.